### PR TITLE
Rename Array -> ArrayReference (towards 363)

### PR DESCRIPTION
**Pull request by arporter**
Purely renaming and splitting out ArrayReference into its own file.

### CODE REVIEW

File: src/psyclone/tests/psyir/nodes/array_reference_test.py [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+from psyclone.psyir.symbols import DataSymbol, ArrayType, \
+    REAL_SINGLE_TYPE, INTEGER_SINGLE_TYPE, REAL_TYPE, INTEGER_TYPE
+from psyclone.errors import GenerationError
+from psyclone.psyir.backend.fortran import FortranWriter
+from psyclone.tests.utilities import check_links
+
+
+def test_array_node_str():
+    ''' Check the node_str method of the ArrayReference class.'''
+    from psyclone.psyir.nodes.node import colored, SCHEDULE_COLOUR_MAP
```
Comment (sergisiso) [inline src/psyclone/tests/psyir/nodes/array_reference_test.py:54]: Can you bring this import to the top-level? Is the only pylint warning in this file.

File: src/psyclone/psyir/nodes/array_reference.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                    len(children), len(symbol.shape)))
+
+        array = ArrayReference(symbol)
+        array.children = children
+        for child in children:
+            child.parent = array
+        return array
+
+    def __str__(self):
+        result = ("ArrayReference" + super(ArrayReference, self).__str__() +
```
Comment (sergisiso) [inline src/psyclone/psyir/nodes/array_reference.py:118]: This currently prints the node name as "ArrayReferenceArrayReference", please remove the first hardcoded string. There are a few tests that needs fixing after this.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter This PR is almost ready, just need to address some issues I commented inline. The tests, pycodestyle, examples and pylint all pass. There is a codecov diff miss in dynamo0p1 but it is not the responsibility of this PR.

File: src/psyclone/psyir/nodes/array_reference.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+from psyclone.psyir.nodes.ranges import Range
+from psyclone.psyir.symbols import DataSymbol, ScalarType
+from psyclone.errors import GenerationError
+from psyclone.psyir.nodes import BinaryOperation, Literal
+
+
+class ArrayReference(Reference):
+    '''
+    Node representing an Array reference. As such it has a reference and a
+    subscript list as children 0 and 1, respectively.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/array_reference.py:53]: Oh yes, fixed.

File: src/psyclone/tests/psyir/nodes/array_reference_test.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/array_reference_test.py:54]: Oops, done.

File: src/psyclone/psyir/nodes/array_reference.py (hunk 2 above)
Comment (arporter) [reply src/psyclone/psyir/nodes/array_reference.py:118]: Oh yes. Embarrassingly, I remember thinking that looked weird when I did the search-n-replace but I didn't do anything about it!

**Comment by arporter** [comment]
Assuming GHA is happy, this is ready for next review.

**Review by sergisiso — APPROVED** [review]
@arporter All issues addressed, this PR is ready to be merged

**Pull request merged**